client/cli/src/runtime.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2017-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use futures::{Future, future, future::FutureExt};
+use futures::select;
+use futures::pin_mut;
+use sc_service::{AbstractService, Configuration};
+use crate::error;
+use crate::informant;
+
+#[cfg(target_family = "unix")]
+async fn main<F, E>(func: F) -> Result<(), Box<dyn std::error::Error>>
+where
+	F: Future<Output = Result<(), E>> + future::FusedFuture,
+	E: 'static + std::error::Error,
+{
+	use tokio::signal::unix::{signal, SignalKind};
+
+	let mut stream_int = signal(SignalKind::interrupt())?;
+	let mut stream_term = signal(SignalKind::terminate())?;
+
+	let t1 = stream_int.recv().fuse();
+	let t2 = stream_term.recv().fuse();
+	let t3 = func;
+
+	pin_mut!(t1, t2, t3);
+
+	select! {
+		_ = t1 => println!("Caught SIGINT"),
+		_ = t2 => println!("Caught SIGTERM"),
+		res = t3 => res?,
+	}
+
+	Ok(())
+}
+
+#[cfg(target_family = "windows")]
```

Comment:
```suggestion
#[cfg(not(unix))]
```

Let's stay as platform-independent as possible.

I'm also not sure if there's a point in catching specifically SIGINT and SIGTERM on Unix rather than using `tokio::signal::ctrlc` everywhere.

Reply:
On Linux background services are usually stopped with SIGTERM (example here with system.d https://www.freedesktop.org/software/systemd/man/systemd.kill.html )

I believe it's best to keep both: SIGINT and SIGTERM for Unix, and ctrlc for the others.

(Btw, on tokio they use: `#[cfg(unix)]` and `#[cfg(windows)]` which are shortcuts for what I did.
https://docs.rs/tokio/0.2.9/src/tokio/signal/ctrl_c.rs.html#50-53 )